### PR TITLE
License Update - MJB - 20230722

### PR DESCRIPTION
**Pull request by Dulux-Oz**
So, I've uploaded this in preparation/anticipation of adding in / fixing up the copyright stuff, etc, so let's not actually pull this until all the info's in, OK?  :-)

Solves #1

### CODE REVIEW

**Comment by haxxxton** [comment]
Do we need the header at the top of each file if the whole repo is covered by it right?

We may be able to automate the addition of it to any output files, but not include it on our working files?

File: README.md
Comment (TheSpyder) [inline README.md]: This isn't normally a part of GPL, but sure 🙂 

```suggestion
Copyright (c) 2023 {{Andrew}}, Andrew Herron, {{Ben}}, Matthew J Black, {{Haxxx}}  & {{James}}.

```

**Review by TheSpyder — DISMISSED** [review]

**Comment by Dulux-Oz** [comment]
@TheSpyder Maybe not, but again, I'm going of my reading from the GPL website...  :-)

**Comment by haxxxton** [comment]
@Dulux-Oz ~~ill fix up the conflicts now~~ Fixed merge conflicts, i amend committed, so you may need to kill and re checkout your branch if you get issues when pulling

File: README.md
Comment (haxxxton) [reply README.md]: updated this when i rebased

**Comment by TheSpyder** [comment]
The mistake was trying to merge the remote into your branch. Haxx did a force push, when that happens you need to throw away your local branch or things get messed up 😆

**Comment by Dulux-Oz** [comment]
Ohhh Kaaay, that piece of information would have been helpful a few hours ago.

I thought we were operating under the general principle that mergers weren't supposed to be done until 2 others had approved it (which, if I'm being honest, is a "rule" I've broken myself: see this merge).

So, for my edification, how are "forced pushes" notified to the rest of us - or is there a particular "everyone knows to do this" process that happens whenever we do (local?) merges, etc?

**Comment by TheSpyder** [comment]
A force push isn't a merge - haxx just waved a magic wand and rewrote your entire branch.

If a PR is active, you'll see lines similar to the above
`haxxxton force-pushed the feature/license_update  branch from b9d0287 to 0c8dc1f 2 hours ago`

If someone tells you they force pushed a branch, or you see the above, the general steps are
* hope you don't have any local changes. If you do, the following will throw them away and we'd need to discuss how to retain them.
* switch your checkout to `main` or `develop`
* delete your copy of the branch
* fetch updates from github
* checkout the branch again

Alternatively, at the command line
```
git fetch
git reset --hard origin/<branchname>
```

But I don't know if the GUI you use supports "hard reset".

**Comment by Dulux-Oz** [comment]
Yeah, well I did have changes, which I knew I'd loose unless I did what I did.

OK, so I'm now wondering about setting up some guidelines around forced pushes - even if it's only "Hey guys, I'm about to force push" - thoughts (from everyone)?

**Comment by Dulux-Oz** [comment]
And yes, GitKracken does support hard resets.  :-)

**Comment by TheSpyder** [comment]
Well in this case, haxx did say
> Fixed merge conflicts, i amend committed, so you may need to kill and re checkout your branch if you get issues when pulling

If you'd prefer, there are git settings to block force pushes.

**Comment by Dulux-Oz** [comment]
Yeah, but that wasn't seen (by me ie the email didn't arrive) until *after* I'd started down the path I ended on :-(

Personally, I'm not comfortable restricting things until/unless we agree we should - the underlying issue, as I see it, is that I'm not aware of the "conventions" around this - that's my issue, yes - but perhaps it might be useful to remember that not everyone is "up to speed" on this stuff - yet!

I don't know... I don't want to be a speed-hump on this or other such issues that may arise, but obviously, something went "wrong" somewhere  :-(

**Comment by TheSpyder** [comment]
I blame git, as much as anything. When you `git pull` the default is to merge remote changes into your branch - which, if someone has done a force push, is _exactly the wrong thing to do_.

I doubt it will be a problem, in general, because we'll all work on our own branches rather than sharing the same branch.

**Comment by Dulux-Oz** [comment]
Which, if I'm understanding things right, should be why we use feature branches for *everything* and no-one commits to dev (or god forbid, main, where we use a release branch - but the same thing) *without* going thru that process (& the resulting "2-man commit approval" - is that right?

**Comment by TheSpyder** [comment]
Yes this was a hopefully unique situation where haxx caused this branch to conflict and updated it while you were still working on it.

**Pull request merged**
